webapi/src/main/scala/org/knora/webapi/slice/admin/repo/service/DefaultObjectAccessPermissionRepoLive.scala

```diff
@@ -40,31 +51,88 @@ final case class DefaultObjectAccessPermissionRepoLive(
-    ZIO.die(UnsupportedOperationException("Mapper not yet fully implemented"))
 }
 
 object DefaultObjectAccessPermissionRepoLive {
   private val mapper = new RdfEntityMapper[DefaultObjectAccessPermission] {
 
-    override def toEntity(resource: RdfResource): IO[RdfError, DefaultObjectAccessPermission] =
-      for {
-        id <- resource.iri.flatMap { iri =>
-                ZIO.fromEither(PermissionIri.from(iri.value).left.map(ConversionError.apply))
-              }
-        forProject <- resource.getObjectIrisConvert[ProjectIri](KnoraAdmin.ForProject).map(_.head)
-      } yield DefaultObjectAccessPermission(id, forProject)
+    private val permissionsDelimiter = '|'
+
+    override def toEntity(resource: RdfResource): IO[RdfError, DefaultObjectAccessPermission] = for {
+      id <- resource.iri.flatMap { iri =>
+              ZIO.fromEither(PermissionIri.from(iri.value).left.map(ConversionError.apply))
+            }
+      forProject          <- resource.getObjectIrisConvert[ProjectIri](KnoraAdmin.ForProject).map(_.head)
+      forGroup            <- resource.getObjectIrisConvert[GroupIri](KnoraAdmin.ForGroup).map(_.headOption)
+      forResourceClass    <- resource.getObjectIrisConvert[InternalIri](KnoraAdmin.ForResourceClass).map(_.headOption)
+      forResourceProperty <- resource.getObjectIrisConvert[InternalIri](KnoraAdmin.ForProperty).map(_.headOption)
+      forWhat <-
+        ZIO.fromEither(ForWhat.from(forGroup, forResourceClass, forResourceProperty)).mapError(ConversionError.apply)
+      permissions <- parsePermissions(resource)
+    } yield DefaultObjectAccessPermission(id, forProject, forWhat, permissions)
+
+    private def parsePermissions(resource: RdfResource) = for {
+      permissionStr     <- resource.getStringLiteralOrFail[String](KnoraBase.HasPermissions)(Right(_))
+      parsedPermissions <- parsePermission(permissionStr)
+    } yield parsedPermissions
+
+    private def parsePermission(permission: String): IO[RdfError, Chunk[DefaultObjectAccessPermissionPart]] =
+      ZIO
+        .foreach(Chunk.fromIterable(permission.split(permissionsDelimiter).map(_.trim))) { token =>
+          token.split(' ') match {
+            case Array(token, groups) => {
+              val part: Either[String, DefaultObjectAccessPermissionPart] =
+                Permission.ObjectAccess
+                  .fromToken(token)
+                  .toRight("No valid Object Access token")
+                  .flatMap { permission =>
+                    Chunk
+                      .fromIterable(groups.split(','))
+                      .map(GroupIri.from)
+                      .foldLeft[Either[String, Chunk[GroupIri]]](Right(Chunk.empty)) {
+                        case (Left(as), Left(a))   => Left(as + "; " + a)
+                        case (Left(as), _)         => Left(as)
+                        case (_, Left(a))          => Left(a)
+                        case (Right(bs), Right(b)) => Right(bs :+ b)
+                      }
```

Comment:
It might be worthwhile to extract this method to reveal a clear signature of the method, which would be more readable than the implementation.

Reply:
I have extracted the method.

=> 74a3227bc29c36e42fe8306ace03504fb9b154a0